File: code/python/dynamic_programming_article-16.py
"""Interval scheduling problem of greedy algorithm"""

from __future__ import annotations
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Run this lesson exercise with positional inputs."""
    return _Solution().eraseOverlapIntervals(*args, **kwargs)


class _Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        n = len(intervals)
        return n - self.intervalSchedule(intervals)
    def intervalSchedule(self, intvs: List[List[int]]) -> int:
        if len(intvs) == 0:
            return 0
        intvs = sorted(intvs, key=lambda x:x[1])
        count = 1
        x_end = intvs[0][1]
        for interval in intvs:
            start = interval[0]
            if start >= x_end:
                count += 1
                x_end = interval[1]
        return count

File: code/javascript/dynamic_programming_article-16.js
/**
 * Interval scheduling problem of greedy algorithm
 */
export function solve(...args) {
  // Entry point for the course exercise: pass inputs directly.
  return eraseOverlapIntervals(...args);
}

var eraseOverlapIntervals = function(intervals) {
    var n = intervals.length;
    return n - intervalSchedule(intervals);
};
var intervalSchedule = function(intvs) {
    if (intvs.length === 0) return 0;
    intvs.sort((a, b) => a[1] - b[1]);
    var count = 1;
    var x_end = intvs[0][1];
    for (var i = 1; i < intvs.length; i++) {
        var start = intvs[i][0];
        if (start >= x_end) {
            count++;
            x_end = intvs[i][1];
        }
    }
    return count;
};
